File: YASAudio/AudioEngine/YASAudioNode.m
//
//  YASAudioNode.m
//  Copyright (c) 2015 Yuki Yasoshima.
//

#import "YASAudioNode.h"
#import "YASAudioEngine.h"
#import "YASAudioConnection.h"
#import "YASMacros.h"
#import "NSException+YASAudio.h"
#import "NSNumber+YASAudio.h"
#import "NSDictionary+YASAudio.h"

@interface YASAudioNodeCore ()

@property (nonatomic, copy) NSDictionary *inputConnectionContainers;
@property (nonatomic, copy) NSDictionary *outputConnectionContainers;

@end

@implementation YASAudioNodeCore

- (void)dealloc
{
    YASRelease(_inputConnectionContainers);
    YASRelease(_outputConnectionContainers);

    _inputConnectionContainers = nil;
    _outputConnectionContainers = nil;

    YASSuperDealloc;
}

- (YASAudioConnection *)inputConnectionForBus:(NSNumber *)bus
{
    YASRaiseIfMainThread;

    YASWeakContainer *container = _inputConnectionContainers[bus];
    return [container autoreleasingObject];
}

- (YASAudioConnection *)outputConnectionForBus:(NSNumber *)bus
{
    YASRaiseIfMainThread;

    YASWeakContainer *container = _outputConnectionContainers[bus];
    return [container autoreleasingObject];
}

- (NSDictionary *)inputConnections
{
    YASRaiseIfMainThread;

    return [_inputConnectionContainers yas_unwrappedDictionaryFromWeakContainers];
}

- (NSDictionary *)outputConnections
{
    YASRaiseIfMainThread;

    return [_outputConnectionContainers yas_unwrappedDictionaryFromWeakContainers];
}

@end

@interface YASAudioNode ()

@property (nonatomic, strong) YASWeakContainer *engineContainer;
@property (atomic, strong) id nodeCore;
@property (atomic, strong) YASAudioTime *renderTime;

@end

@implementation YASAudioNode {
    NSMutableDictionary *_inputConnectionContainers;
    NSMutableDictionary *_outputConnectionContainers;
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        _inputConnectionContainers = [[NSMutableDictionary alloc] initWithCapacity:8];
        _outputConnectionContainers = [[NSMutableDictionary alloc] initWithCapacity:8];
    }
    return self;
}

- (void)dealloc
{
    @autoreleasepool
    {
        for (YASWeakContainer *container in _inputConnectionContainers.objectEnumerator) {
            YASAudioConnection *inputConnection = [container retainedObject];
            [inputConnection removeDestinationNode];
            YASRelease(inputConnection);
        }
        for (YASWeakContainer *container in _outputConnectionContainers.objectEnumerator) {
            YASAudioConnection *outputConnection = [container retainedObject];
            [outputConnection removeSourceNode];
            YASRelease(outputConnection);
        }
    }

    YASRelease(_inputConnectionContainers);
    YASRelease(_outputConnectionContainers);
    YASRelease(_renderTime);
    YASRelease(_nodeCore);
    YASRelease(_engineContainer);

    _inputConnectionContainers = nil;
    _outputConnectionContainers = nil;
    _renderTime = nil;
    _nodeCore = nil;
    _engineContainer = nil;

    YASSuperDealloc;
}

- (void)reset
{
    [_inputConnectionContainers removeAllObjects];
    [_outputConnectionContainers removeAllObjects];

    [self updateNodeCore];
}

- (void)updateConnections
{
}

- (YASAudioFormat *)inputFormatForBus:(NSNumber *)bus
{
    YASRaiseIfSubThread;

    YASWeakContainer *container = [_inputConnectionContainers objectForKey:bus];
    YASAudioConnection *connection = [container retainedObject];
    YASAudioFormat *format = connection.format;
    YASRelease(connection);
    return format;
}

- (YASAudioFormat *)outputFormatForBus:(NSNumber *)bus
{
    YASRaiseIfSubThread;

    YASWeakContainer *container = [_outputConnectionContainers objectForKey:bus];
    YASAudioConnection *connection = [container retainedObject];
    YASAudioFormat *format = connection.format;
    YASRelease(connection);
    return format;
}

- (void)setEngine:(YASAudioEngine *)engine
{
    self.engineContainer = engine.weakContainer;
}

- (YASAudioEngine *)engine
{
    return [_engineContainer autoreleasingObject];
}

- (YASAudioTime *)lastRenderTime
{
    return self.renderTime;
}

#pragma mark Core

- (Class)nodeCoreClass
{
    return [YASAudioNodeCore class];
}

- (id)newNodeCoreObject
{
    YASAudioNodeCore *core = [[[self nodeCoreClass] alloc] init];
    core.inputConnectionContainers = _inputConnectionContainers;
    core.outputConnectionContainers = _outputConnectionContainers;

    return core;
}

- (void)updateNodeCore
{
    id core = [self newNodeCoreObject];
    self.nodeCore = core;
    YASRelease(core);
}

#pragma mark Connection

- (void)addConnection:(YASAudioConnection *)connection
{
    if (connection.destinationNode == self) {
        _inputConnectionContainers[connection.destinationBus] = connection.weakContainer;
    } else if (connection.sourceNode == self) {
        _outputConnectionContainers[connection.sourceBus] = connection.weakContainer;
    } else {
        YASRaiseWithReason(
            ([NSString stringWithFormat:@"%s - Connection does not exist in a node.", __PRETTY_FUNCTION__]));
    }

    [self updateNodeCore];
}

- (void)removeConnection:(YASAudioConnection *)connection
{
    if (connection.destinationNode == self) {
        NSNumber *toBus = connection.destinationBus;
        [_inputConnectionContainers removeObjectForKey:toBus];
    }

    if (connection.sourceNode == self) {
        NSNumber *fromBus = connection.sourceBus;
        [_outputConnectionContainers removeObjectForKey:fromBus];
    }

    [self updateNodeCore];
}

- (YASAudioConnection *)inputConnectionForBus:(NSNumber *)bus
{
    YASRaiseIfSubThread;

    YASWeakContainer *container = _inputConnectionContainers[bus];
    return [container autoreleasingObject];
}

- (YASAudioConnection *)outputConnectionForBus:(NSNumber *)bus
{
    YASRaiseIfSubThread;

    YASWeakContainer *container = _outputConnectionContainers[bus];
    return [container autoreleasingObject];
}

- (NSDictionary *)inputConnections
{
    YASRaiseIfSubThread;

    return [_inputConnectionContainers yas_unwrappedDictionaryFromWeakContainers];
}

- (NSDictionary *)outputConnections
{
    YASRaiseIfSubThread;

    return [_outputConnectionContainers yas_unwrappedDictionaryFromWeakContainers];
}

#pragma mark Bus

- (UInt32)inputBusCount
{
    return 0;
}

- (UInt32)outputBusCount
{
    return 0;
}

- (NSNumber *)nextAvailableInputBus
{
    return [_inputConnectionContainers yas_emptyNumberKeyInLength:self.inputBusCount];
}

- (NSNumber *)nextAvailableOutputBus
{
    return [_outputConnectionContainers yas_emptyNumberKeyInLength:self.outputBusCount];
}

- (BOOL)isAvailableInputBus:(NSNumber *)bus
{
    if (bus.uint32Value >= self.inputBusCount) {
        return NO;
    }

    return _inputConnectionContainers[bus] == nil;
}

- (BOOL)isAvailableOutputBus:(NSNumber *)bus
{
    if (bus.uint32Value >= self.outputBusCount) {
        return NO;
    }

    return _outputConnectionContainers[bus] == nil;
}

#pragma mark Render thread

- (void)renderWithData:(YASAudioData *)data bus:(NSNumber *)bus when:(YASAudioTime *)when
{
    [self setRenderTimeOnRender:when];
}

- (void)setRenderTimeOnRender:(YASAudioTime *)time
{
    self.renderTime = time;
}

@end

File: YASAudio/AudioEngine/yas_audio_node.cpp
//
//  yas_audio_node.cpp
//  Copyright (c) 2015 Yuki Yasoshima.
//

#include "yas_audio_node.h"
#include "yas_audio_engine.h"
#include "yas_audio_connection.h"
#include "yas_audio_time.h"

using namespace yas;

audio_node::audio_node(std::nullptr_t) : super_class(nullptr)
{
}

audio_node::audio_node(const std::shared_ptr<audio_node::impl> &impl) : super_class(impl)
{
    if (impl) {
        impl->set_node(*this);
    }
}

audio_node::~audio_node() = default;

void audio_node::reset()
{
    if (!impl_ptr()) {
        std::cout << "_impl is null" << std::endl;
    }
    _impl_ptr()->reset();
}

audio_format audio_node::input_format(const UInt32 bus_idx) const
{
    return _impl_ptr()->input_format(bus_idx);
}

audio_format audio_node::output_format(const UInt32 bus_idx) const
{
    return _impl_ptr()->output_format(bus_idx);
}

bus_result_t audio_node::next_available_input_bus() const
{
    return _impl_ptr()->next_available_input_bus();
}

bus_result_t audio_node::next_available_output_bus() const
{
    return _impl_ptr()->next_available_output_bus();
}

bool audio_node::is_available_input_bus(const UInt32 bus_idx) const
{
    return _impl_ptr()->is_available_input_bus(bus_idx);
}

bool audio_node::is_available_output_bus(const UInt32 bus_idx) const
{
    return _impl_ptr()->is_available_output_bus(bus_idx);
}

audio_engine audio_node::engine() const
{
    return _impl_ptr()->engine();
}

audio_time audio_node::last_render_time() const
{
    return _impl_ptr()->render_time();
}

UInt32 audio_node::input_bus_count() const
{
    return _impl_ptr()->input_bus_count();
}

UInt32 audio_node::output_bus_count() const
{
    return _impl_ptr()->output_bus_count();
}

#pragma mark render thread

void audio_node::render(audio_pcm_buffer &buffer, const UInt32 bus_idx, const audio_time &when)
{
    _impl_ptr()->render(buffer, bus_idx, when);
}

void audio_node::set_render_time_on_render(const audio_time &time)
{
    _impl_ptr()->set_render_time_on_render(time);
}

#pragma mark - private

std::shared_ptr<audio_node::impl> audio_node::_impl_ptr() const
{
    return impl_ptr<impl>();
}

audio_connection audio_node::_input_connection(const UInt32 bus_idx) const
{
    return _impl_ptr()->input_connection(bus_idx);
}

audio_connection audio_node::_output_connection(const UInt32 bus_idx) const
{
    return _impl_ptr()->output_connection(bus_idx);
}

const audio_connection_wmap &audio_node::_input_connections() const
{
    return _impl_ptr()->input_connections();
}

const audio_connection_wmap &audio_node::_output_connections() const
{
    return _impl_ptr()->output_connections();
}

void audio_node::_add_connection(const audio_connection &connection)
{
    _impl_ptr()->add_connection(connection);
}

void audio_node::_remove_connection(const audio_connection &connection)
{
    _impl_ptr()->remove_connection(connection);
}

void audio_node::_set_engine(const audio_engine &engine)
{
    _impl_ptr()->set_engine(engine);
}

audio_engine audio_node::_engine()
{
    return _impl_ptr()->engine();
}

void audio_node::_update_kernel()
{
    _impl_ptr()->update_kernel();
}

void audio_node::_update_connections()
{
    _impl_ptr()->update_connections();
}
